singularity-eos/eos/eos_base.hpp

```diff
@@ -0,0 +1,314 @@
+    dpdr = (Pr - press) / drho;
+    dpde = (Pe - press) / de;
+    dtdr = (Tr - temp) / drho;
+    dtde = (Te - temp) /
+           de; // Would it be better to skip the calculation of Te and return 1/cv?
+    return;
+  }
+  // Specialzied vector version of PTofRE maybe more suited for EOSPAC
+  // }
+  // template<typename RealIndexer, typename LambdaIndexer>
+  // inline
+  // void PTofRE(RealIndexer &&rhos, RealIndexer &&sies,
+  //             RealIndexer &&presses, RealIndexer &&temps,
+  //             RealIndexer &&dpdrs, RealIndexer &&dpdes,
+  //             RealIndexer &&dtdrs, RealIndexer &&dtdes,
+  //             const int num, LambdaIndexer &&lambdas) const {
+  //   // Get the dervived class
+  //   auto eos = static_cast<CRTP const&>(*this);
+
+  //   // Lookup at density and energy
+  //   eos.PressureFromDensityInternalEnergy(rhos, sies, num, presses,
+  //                                               lambdas);
+  //   eos.TemperatureFromDensityInternalEnergy(rhos, sies, num, temps,
+  //                                                  lambdas);
+  //   // Peturbation factor
+  //   Real factor = 1. + 1.0e-06;
+
+  //   // Perturb densities and do lookups
+  //   portableFor(
+  //       'PerturbDensities', 0, num, PORTABLE_LAMBDA(const int i) {
+  //         rhos[i] *= factor;
+  //       }
+  //   );
+  //   eos.PressureFromDensityInternalEnergy(rhos, sies, num, dpdrs,
+  //                                               lambdas);
+  //   eos.TemperatureFromDensityInternalEnergy(rhos, sies, num, dtdrs,
+  //                                                  lambdas);
+
+  //   // Reset densities, perturb energies, and do lookups
+  //   portableFor(
+  //       'PerturbEnergiesResetDensities', 0, num,
+  //       PORTABLE_LAMBDA(const int i) {
+  //         sies[i] *= factor;
+  //         rhos[i] /= factor;
+  //       }
+  //   );
+  //   eos.PressureFromDensityInternalEnergy(rhos, sies, dpdes, num,
+  //                                               lambdas);
+  //   eos.TemperatureFromDensityInternalEnergy(rhos, sies, dtdes, num,
+  //                                                  lambdas);
+
+  //   // Reset the energies to their original values
+  //   portableFor(
+  //       'ResetEnergies', 0, num, PORTABLE_LAMBDA(const int i) {
+  //         sies[i] /= factor;
+  //       }
+  //   );
+
+  //   // Calculate the derivatives
+  //   portableFor(
+  //       'CalculateDerivatives', 0., num, PORTABLE_LAMBDA(const int i) {
+  //         dpdrs[i] = (dpdrs[i] - presses[i]) / (rhos[i] * (1. - factor));
+  //         dpdes[i] = (dpdes[i] - presses[i]) / (sies[i] * (1. - factor));
+  //         dtdrs[i] = (dtdrs[i] - temps[i]) / (rhos[i] * (1. - factor));
+  //         dtdes[i] = (dtdes[i] - temps[i]) / (sies[i] * (1. - factor));
+  //       }
+  //   );
+
+  //   return;
+  // }
+};
+} // namespace eos_base
+} // namespace singularity
+
```

Comment:
not necessary but could add an `#undef SG_MEMBER_FUNC_NAME` for completeness.

Reply:
I am in favor of this.

Reply:
Sure. Added a `#undef SG_ADD_BASE_CLASS_USINGS` as well.

Reply:
I don't think you can `#undef` `SG_ADD_BASE_CLASS_USINGS` because it's needed in `eos.hpp` and other header files right?

Reply:
I assumed the @dholladay00 meant adding an `#undef` right before the `#define` so I just did the same thing for `SG_ADD_BASE_CLASS_USINGS`. Did I misunderstand?

The `#undef` should just make ensure there aren't any name conflicts right?

Reply:
I'll assume that @dholladay00 wanted the `#undef` to go at the end of the file then and I'll remove both of these since they should probably be wrapped by `#ifdef`.

Reply:
Done